pkg/sql/logictest/testdata/logic_test/composite_types

```diff
@@ -0,0 +1,213 @@
+statement ok
+CREATE TYPE t AS (a INT, b INT)
+
+statement ok
+DROP TYPE t
+
+statement ok
+CREATE TYPE t AS (a INT, b INT)
+
+statement error pq: relation "t" does not exist
+SELECT * FROM t
+
+statement error pq: type "test.public.t" already exists
+CREATE TABLE t (x INT)
+
+statement error pq: type "test.public.t" already exists
+CREATE TYPE t AS (a INT)
+
+statement ok
+CREATE TABLE torename (x INT)
+
+statement error pq: type "test.public.t" already exists
+ALTER TABLE torename RENAME TO t
+
+query TII
+SELECT (1, 2)::t, ((1, 2)::t).a, ((1, 2)::t).b
+----
+(1,2)  1  2
+
+statement error could not identify column \"foo\"
+SELECT ((1, 2)::t).foo
+
+statement ok
+CREATE TABLE tab (a t)
+
+statement ok
+INSERT INTO tab VALUES (NULL), ((1, 2))
+
+statement ok
+INSERT INTO tab VALUES ((1, NULL))
+
+# TODO(jordan): this should work fine.
```

Comment:
Isn't this a rather serious correctness bug? Seems like a showstopper to me at least.

Reply:
It's not a correctness bug the way we define it - errors where there shouldn't be errors are not correctness bugs.

The reason this is broken is a type system issue that's unrelated to composite types. I filed #93349 for it but I don't agree that it's a showstopper - it's a classic example of CockroachDB's type system differences from Postgres, and is correctable by adding casts. I've improved the comments here.

Here's an example of the existing broken-ness:

```
demo@127.0.0.1:26257/defaultdb> values((1, 2)), ((1, null));

ERROR: VALUES types tuple{int, unknown} and tuple{int, int} cannot be matched
SQLSTATE: 42804
```